keps/sig-node/20191211-node-maintenance-lease.md

```diff
@@ -0,0 +1,175 @@
+another controller or user is performing a disruptive operation against a node.
+Examples of disruptive operations include:
+1. Rebooting a node
+1. Powering off a node for maintenance
+1. Draining or Cordoning a node
+1. Deleting/deprovisioning a node due to failed health checks
+
+### Goals
+
+1. Define a central point of coordination for various actors.
+1. Define high-level steps for well-behaved clients to implement.
+
+### Non-Goals
+
+Define actors which might utilize the lease.
+
+## Proposal
+
+Utilize the existing `Lease` built-in API in the API group `coordination.k8s.io`.
+Create a new Lease object per-node  with Name equal to Node name in a newly
+created dedicated namespace
+That namespace should be created automatically (similarly to "default" and
+"kube-system", probably by NodeController) and never be deleted (so that nodes
+don't require permission for it).
+
+The Lease object should be created automatically, and the ownerRef should
+be the corresponding node so it is removed when the node is deleted.
+
+### User Stories [optional]
+
+#### Story 1
+
+I have a configuration management system that requires the node to be online to perform configuration updates.
+
+I have an automated-snapshot system that stops an instance and takes a snapshot for backup / data retention purposes.
+
+I do not want snapshots to happen during an update.
+
+#### Story 2
+
+I am investigating a problem with the node. I don't want the node to be disrupted
+by automation while I'm performing my investigation or fixes, and I don't want
+to stop the automation components from acting on other nodes.
+
+### Implementation Details/Notes/Constraints
+
+#### Impact to the Kubelet
+
+None. The kubelet does not need to respond to this lease or interact with it in
+any way.  This is for other components to coordination actions they intend to
+take (or prevent) on the node that might be disruptive.
+
+#### Impact to Scheduleability
+
+None. The existence of acquiring of this lease does not imply any taints,
+annotations, labels, cordoning or draining will take place. The Lease object
+itself is not intended as a mechanism to start any automated process.
+
+A user/controller acquires the Lease object, and then that user/controller
+may add/remove taints, cordon or drain, if they choose to.  Other components
+may wish to integrate with the new Lease object to avoid conflict. EG:
+a component that ensures a taint is always present on a node might stop ensuring
+that taint if there is a current valid lease holder, but that implementation
+is out of scope here.
+
+#### Client Behavior
+
+1. Client has required RBAC permissions for new namespace and Lease object
+1. Client checks for the existence of a lease (to be backwards compatible)
+1. If the lease exists, confirm the lease is not currently held by another user
+1. Attempt to acquire the lease and set an appropriate LeaseDurationSeconds
+1. If acquired, perform the necessary operations
+1. Release the lease
```

Comment:
2 Things,

This is a list of 1s, should make sure to number them before merging

The leaseDurationSeconds, is this a field on the lease object? I'm not familiar with them (perhaps a link where they are first mentioned may be helpful to readers) Is it assumed that if this duration has been exceeded then anyone is free to attempt to acquire the lease? What prevents a client from overwriting the lease? Is that built in or something we need to consider?

Reply:
List of 1's generates a numbered list, it's markdown syntax.

Reply:
> List of 1's generates a numbered list, it's markdown syntax.

Yep good point 🤦  Second half of the comment is still prevalent though I think